### ISSUE TITLE
Max amount of fixtures

### ISSUE THREAD

**Issue by darthf1** [hautelook/AliceBundle#330]
Hi. I'm currently on `2.0@beta` and trying to load 100k database records through Alice. When i load 1.000 records, it's going within a second. When i load 10.000 fixtures, i'm given the warning that the database will be purged but after that, nothing really happens. 

Is Alice/ this bundle meant for such an amount of fixtures? Or am i just doing it wrong? Is there a way to maybe load 10x 1.000 fixtures, instead of 1x 10.000 (without copy pasting the fixtures 10 times)?

**Comment by theofidry** [comment]
Interesting question. I would say it highly depends on your fixtures. Alice by default is quite generic, and it allows you do generate a certain number of objects and then persist them. There a few potential bottlenecks:

- The sheer number of data to generate. I don't know how big are the data sets of Faker, but I'm not sure it can generate 100K fixtures without a lot of redundancy.
- The sheer number of objects: Alice have to load everything at once. It is not designed to receive a stream of files to load and output a stream of objects, simply because there might be dependencies between objects so this kind of architecture would be a complete nightmare to design.
- The persistence layer: Doctrine ORM for example (but that's true for any ORM), does pretty badly at persisting huge amounts of objects.

The best use case for Alice I think is to load small amounts which fits specific scenarios you may need for integration tests, or enough data for a demo or something. If you want to load huge amounts of data, it is doable, but you'll have to dig in the library, identity the bottleneck and see how to solve them. There is actually a lot of places that can be optimized but are not done by default in Alice simply because it depends too much on your case and demands to know too much of your domain. But here's a few leads:

- Being careful of [Alice DSL](https://github.com/nelmio/alice/blob/master/doc/advanced-guide.md#expression-language-dsl). For example `@user*` requires alice to retrieve all the IDs and filter them to know which one are valid.
- Watch of the `unique` flag. It is a very nice feature but a costly operation. Basically the idea is that a pool stores the values being used, have a max number of try, and when a value is being resolved/use, store it in the pool. If already there, will try again until obtaining a new value or reaching the number of try. Simple, but dreadful if you have too few unique data upstream.
- There is a few places where Alice tries to instantiate, access to some values or set values in a generic way. While that works (and fast enough in most cases), it is certainly not as performant as you might want and you could replace that part by something more specific.
- How you are loading files: by default, Alice tries to load all the files first and then persist them. As a result, all the objects are loaded into memory before being persisted, and they are all persisted at once... In your case I think you use a different strategy to be more efficient. It's not done by default because of potential dependencies between files, which Alice doesn't know about.
- How you are persisting your objects: you could rely on a custom persister instead of Doctrine or another ORM to persist your objects in a more efficient manner.

TL:DR; Alice is not really meant for that. It's doable, but you'll have to dig into it.

**Comment by darthf1** [comment]
Thanks for your response. Indeed I see Alice as a way to populate your database with realistic data. What im trying to do is pump a number of records in the database, to test various interfaces. Because later in my project i need the realistic data, i was trying to do it with Alice so i have one solution for all.

For now, for this, im going with DoctrineFixtures and just persist 100k rows, and flushing every 1000 rows. 

Thanks again!
